### PR TITLE
Ea 2891 new method enrichment

### PR DESCRIPTION
**Pull request by SrishtiSingh-eu**

### CODE REVIEW

File: entity-web/src/main/java/eu/europeana/entity/web/controller/BaseRest.java [hunk 1]
```diff
@@ -228,6 +227,27 @@ protected String extractSolrFieldName(String query) {
...
+	protected void validateLanguage(String language) throws ParamValidationException {
+	if (StringUtils.isEmpty(language))	{
+		return;
+	}
+	if (StringUtils.contains(language, WebEntityConstants.COMMA)) {
+		throw new ParamValidationException(I18nConstants.UNSUPPORTED_MULTIPLE_LANG_VALUE,
+				CommonApiConstants.QUERY_PARAM_LANG, language);
+	}
+
+	if (!language.matches(WebEntityConstants.LANG_REGEX)) {
```
Comment (gsergiu) [inline entity-web/src/main/java/eu/europeana/entity/web/controller/BaseRest.java:244]: if we use the 2 letter codes only, we could use the Locale.getISOLanguages() to validate the request parameter

**Review by gsergiu — COMMENTED** [review]
there are a few minor updates required and we need to clarify if the support for 3 letter codes is needed (there is currently no data for 3 letter codes)

File: entity-definitions/src/main/java/eu/europeana/entity/definitions/model/vocabulary/WebEntityConstants.java [hunk 2]
```diff
@@ -44,6 +50,7 @@ public interface WebEntityConstants extends WebEntityFields {
     public static final String FIELD_DELIMITER = ":";
     public static final String SOLR_AND = " AND ";
     public static final String SOLR_OR = " OR ";
+    public static final String SOLR_DESC = " desc";
```
[the remark this answers is not in the document]
Comment (SrishtiSingh-eu) [reply entity-definitions/src/main/java/eu/europeana/entity/definitions/model/vocabulary/WebEntityConstants.java:53]: done

File: entity-definitions/src/main/java/eu/europeana/entity/definitions/model/vocabulary/WebEntityFields.java [hunk 3]
```diff
@@ -23,6 +23,10 @@ public interface WebEntityFields {
     public static final String THUMBNAIL = "thumbnail";
     public static final String IS_SHOWN_BY = "isShownBy";
 
+    //Enrichment field
+    public static final String ENRICH_LABEL_FIELD = "label_enrich";
+    public static final String DERIVED_SCORE = "derived_score";
```
[the remark this answers is not in the document]
Comment (SrishtiSingh-eu) [reply entity-definitions/src/main/java/eu/europeana/entity/definitions/model/vocabulary/WebEntityFields.java:28]: done

File: entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java [hunk 4]
```diff
@@ -314,6 +314,46 @@ public Query buildSearchQuery(String queryString, String[] filters, int pageSize
...
+	private String[] createFilterForEnrichment(List<EntityTypes> entityTypes) {
+		String entityFilter = buildEntityTypeCondition(entityTypes);
+		if(entityFilter != null) {
+			return toArray(WebEntityConstants.TYPE + WebEntityConstants.FIELD_DELIMITER + entityFilter);
+		}
+		return new String[0];
+	}
+
+	private String[] createSortForEnrichment(){
+    	return toArray(WebEntityConstants.DERIVED_SCORE + WebEntityConstants.SOLR_DESC);
```
[the remark this answers is not in the document]
Comment (SrishtiSingh-eu) [reply entity-solr/src/main/java/eu/europeana/entity/solr/service/impl/EntityQueryBuilder.java:354]: done

File: entity-web/src/main/java/eu/europeana/entity/web/controller/BaseRest.java [hunk 5]
```diff
@@ -228,6 +227,27 @@ protected String extractSolrFieldName(String query) {
 	return fieldName;
     }
 
+	/**
+	 * Validate language parameter
+	 * @param language
+	 * @throws ParamValidationException
+	 */
+	protected void validateLanguage(String language) throws ParamValidationException {
```
[the remark this answers is not in the document]
Comment (SrishtiSingh-eu) [reply entity-web/src/main/java/eu/europeana/entity/web/controller/BaseRest.java:235]: I have adjusted the validation. Enrichment query is created in -> **EntityQueryBuilder.createSearchQueryForEnrichment(**) method.

File: entity-web/src/main/java/eu/europeana/entity/web/controller/BaseRest.java (hunk 1 above)
Comment (SrishtiSingh-eu) [reply entity-web/src/main/java/eu/europeana/entity/web/controller/BaseRest.java:244]: done

File: entity-definitions/src/main/java/eu/europeana/entity/definitions/model/vocabulary/WebEntityFields.java [hunk 6]
```diff
@@ -23,6 +23,9 @@ public interface WebEntityFields {
     public static final String THUMBNAIL = "thumbnail";
     public static final String IS_SHOWN_BY = "isShownBy";
 
+    //Enrichment field
+    public static final String ENRICH_LABEL_FIELD = "label";
```
Comment (gsergiu) [inline entity-definitions/src/main/java/eu/europeana/entity/definitions/model/vocabulary/WebEntityFields.java:27]: shouldn't this be label_enrich?

File: entity-web/src/main/java/eu/europeana/entity/web/controller/BaseRest.java [hunk 7]
```diff
@@ -228,6 +229,28 @@ protected String extractSolrFieldName(String query) {
...
+		return;
+	}
+	// multiple language not supported
+	if (StringUtils.contains(language, WebEntityConstants.COMMA)) {
+		throw new ParamValidationException(I18nConstants.UNSUPPORTED_MULTIPLE_LANG_VALUE,
+				CommonApiConstants.QUERY_PARAM_LANG, language);
+	}
+	// language value can be 'all' Or ISO language only
+	if (!StringUtils.equals(language, WebEntityConstants.PARAM_LANGUAGE_ALL) &&
+	     Arrays.stream(Locale.getISOLanguages()).noneMatch(isoLang -> StringUtils.equals(language, isoLang))){
```
Comment (gsergiu) [inline entity-web/src/main/java/eu/europeana/entity/web/controller/BaseRest.java:248]: I think we should improve the performance by implementing validation like in Entity Management
By using a static set and use the contains method
see:
 private static final Set<String> ISO_LANGUAGES = Set.of(Locale.getISOLanguages());

https://github.com/europeana/entity-management/blob/6deb5764b59f0a481547ec0d8ca10ca6fd8d3b3a/entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsCleaner.java#L309

**Review by gsergiu — CHANGES_REQUESTED** [review]

File: entity-definitions/src/main/java/eu/europeana/entity/definitions/model/vocabulary/WebEntityFields.java (hunk 6 above)
Comment (SrishtiSingh-eu) [reply entity-definitions/src/main/java/eu/europeana/entity/definitions/model/vocabulary/WebEntityFields.java:27]: oh yes, i think for testing i changed it and for got to revert it.

File: entity-web/src/main/java/eu/europeana/entity/web/controller/BaseRest.java (hunk 7 above)
Comment (SrishtiSingh-eu) [reply entity-web/src/main/java/eu/europeana/entity/web/controller/BaseRest.java:248]: done

**Pull request merged**